java/toolchains/BUILD

```diff
@@ -0,0 +1,92 @@
+load("@rules_java//java:defs.bzl", "java_toolchain")
+
+toolchain_type(name = "javac_toolchain_type")
+toolchain_type(name = "java_runtime_toolchain_type")
+
+LINUX_CONSTRAINTS = ["@platforms//os:linux", "@platforms//cpu:x86_64",]
+WINDOWS_CONSTRAINTS = ["@platforms//os:windows", "@platforms//cpu:x86_64",]
+MACOS_CONSTRAINTS = ["@platforms//os:osx", "@platforms//cpu:x86_64",]
+
+HOST_JAVA_RUNTIME_CONSTRAINTS = ["@rules_java//java/constraints:host_runtime"]
+HOST_JAVA_LANGUAGE_CONSTRAINTS = ["@rules_java//java/constraints:host_language"]
+
+############################### javac toolchains ##############################
+toolchain(
+    name = "javac_linux_toolchain",
+    exec_compatible_with = LINUX_CONSTRAINTS,
+    target_compatible_with = LINUX_CONSTRAINTS,
+    toolchain = "@remote_java_tools_linux//:toolchain",
+    toolchain_type = ":javac_toolchain_type"
+)
+
+toolchain(
+    name = "javac_windows_toolchain",
+    exec_compatible_with = WINDOWS_CONSTRAINTS,
+    target_compatible_with = WINDOWS_CONSTRAINTS,
+    toolchain = "@remote_java_tools_windows//:toolchain",
+    toolchain_type = ":javac_toolchain_type"
+)
+
+toolchain(
+    name = "javac_darwin_toolchain",
+    exec_compatible_with = MACOS_CONSTRAINTS,
+    target_compatible_with = MACOS_CONSTRAINTS,
+    toolchain = "@remote_java_tools_darwin//:toolchain",
+    toolchain_type = ":javac_toolchain_type",
+)
+
+toolchain(
+    name = "host_javac_linux_toolchain",
+    exec_compatible_with = LINUX_CONSTRAINTS + HOST_JAVA_LANGUAGE_CONSTRAINTS,
+    target_compatible_with = LINUX_CONSTRAINTS + HOST_JAVA_LANGUAGE_CONSTRAINTS,
+    toolchain = "@remote_java_tools_linux//:toolchain",
+    toolchain_type = ":javac_toolchain_type"
+)
+
+toolchain(
+    name = "host_javac_windows_toolchain",
+    exec_compatible_with = WINDOWS_CONSTRAINTS + HOST_JAVA_LANGUAGE_CONSTRAINTS,
+   target_compatible_with = WINDOWS_CONSTRAINTS + HOST_JAVA_LANGUAGE_CONSTRAINTS,
+   toolchain = "@remote_java_tools_windows//:toolchain",
+   toolchain_type = ":javac_toolchain_type"
+)
+
+toolchain(
+    name = "host_javac_darwin_toolchain",
+    exec_compatible_with = MACOS_CONSTRAINTS + HOST_JAVA_LANGUAGE_CONSTRAINTS,
+    target_compatible_with = MACOS_CONSTRAINTS + HOST_JAVA_LANGUAGE_CONSTRAINTS,
+    toolchain = "@remote_java_tools_darwin//:toolchain",
+    toolchain_type = ":javac_toolchain_type",
+)
+
+############################ java runtime toolchains ##########################
+
+toolchain(
+    name = "java_runtime_toolchain",
+    toolchain = "@local_jdk//:jdk",
+    toolchain_type = ":java_runtime_toolchain_type"
+)
+
+toolchain(
+    name = "host_java_runtime_linux_toolchain",
+    exec_compatible_with = LINUX_CONSTRAINTS + HOST_JAVA_RUNTIME_CONSTRAINTS,
+    target_compatible_with = LINUX_CONSTRAINTS + HOST_JAVA_RUNTIME_CONSTRAINTS,
+    toolchain = "@host_jdk_linux//:jdk",
```

Comment:
Here again, if we register java_runtime_toolchain before host_java_runtime_linux_toolchain, it will always be selected. Please correct my understanding if I'm wrong. I assume you wanted to give people the sane default of having java_runtime_toolchain available, and then give them a way to select downloaded java runtime when they want to use that or when they don't have the local_jdk installed at all?

Since user has to be explicit in their WORKSPACE file, we can tell them to either register java_runtime_toolchain, or host_java_runtime_linux_toolchain. That would be one way of choosing which one to use.

Alternative would be to actaully detect what java runtime is installed on the machine (I'd even inspect its version), and put exec_compatible_with constraint to java_runtime_toolchain, and also in the host platform. That tells Bazel that my host machine fulfills this constraints, and that java_runtime_toolchain can be selected. This way we can register both java_runtime_toolchain and host_java_runtime_linux_toolchain. 

Does this make any sense?

Reply:
I think the terminology is a bit confusing. We previously had target and host tools. We now have target and execution platforms. The toolchains defined here should preserve the same state that Bazel currently defaults to:
- `@local_jdk//:jdk` for the target configuration
- Our remote jdk with a specific java version (11) `@host_jdk_linux//:jdk` for the host configuration

If the users need different toolchains they have to define and register them themselves.

After defining & registering these toolchains, my plan was to use

 ```
--platforms=@local_config_platform//:host \
--host_platform=@rules_java//java/platforms:host_java_platform
```

which should select `@rules_java//java/toolchains:java_runtime_toolchain` for the target config and `@rules_java//java/toolchains:host_java_runtime_X_toolchain` for the host config. From your comment I understand that this won't happen, even if I specify a host_platform? I would like to find a solution where we can register both toolchains.

Reply:
Part of the output generated by `--toolchain_resolution_debug`:

```
Selected execution platform @rules_java//java/platforms:host_java_platform, type @rules_java//java/toolchains:java_runtime_toolchain_type -> toolchain @local_jdk//:jdk
Selected execution platform @rules_java//java/platforms:host_java_platform, type @rules_java//java/toolchains:java_runtime_toolchain_type -> toolchain @host_jdk_linux//:jdk
Selected execution platform @rules_java//java/platforms:host_java_platform, type @rules_java//java/toolchains:javac_toolchain_type -> toolchain @remote_java_tools_linux//:toolchain
```

which confuses mesurther, since it doesn't seem to try to find toolchains for the target platform, but it selects 2 runtime toolchains for the host platform. So I'm not understanding something. :)

Reply:
I see. There is only one execution platform selected. The problem is that for Java we have the same target & host platform, but we need use another java version for building the tools in the host configuration. I can define the host toolchain using an alias and with no os/cpu constraints, e.g.:

```
toolchain(
    name = "host_javac_toolchain",
    exec_compatible_with = HOST_JAVA_LANGUAGE_CONSTRAINTS,
    target_compatible_with = HOST_JAVA_LANGUAGE_CONSTRAINTS,
    toolchain = select({
        "@bazel_tools//src/conditions:darwin": "@remote_java_tools_darwin//:toolchain",
        "@bazel_tools//src/conditions:windows": "@remote_java_tools_windows//:toolchain",
        "@bazel_tools//src/conditions:linux_x86_64": "@remote_java_tools_linux//:toolchain",
    }),
    toolchain_type = ":javac_toolchain_type"
)
```

Will try this after lunch.

Reply:
Obsolete.